orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -706,13 +827,39 @@ Status ToCheckpointState(gsl::span<const uint8_t> checkpoint_bytes, CheckpointSt
                     "Checkpoint is invalid. Checkpoint version ", checkpoint_version, " is not supported.");
 
   const auto* fbs_module_state = fbs_checkpoint->module_state();
+
+  fbs::utils::ExternalDataReader external_data_reader = nullptr;
+  std::optional<std::ifstream> external_data_stream;
+
+  if (fbs_module_state->has_external_data()) {
+    ORT_RETURN_IF_NOT(checkpoint_path.has_value(),
+                      "External data is present in the checkpoint but the checkpoint path is not provided. External data with loading from buffer is not supported yet.");
+    auto data_path = ExternalCheckpointDataPath(checkpoint_path.value());
+    external_data_stream = std::ifstream(data_path, std::ios::binary);
+
+    char errbuf[256];
+    ORT_RETURN_IF(external_data_stream.value().fail(),
+                  "Failed to open checkpoint's external data file: ", ToUTF8String(data_path), strerror_s(errbuf, sizeof(errbuf), errno));
```

Comment:
should we use strerrorlen_s to query the required buffer length? or does directly calling strerror work?

Reply:
Ahh I had tried with strerror originally, and I got the error `error C4996: 'strerror': This function or variable may be unsafe. Consider using strerror_s instead. To disable deprecation, use _CRT_S
ECURE_NO_WARNINGS. See online help for details.` when calling strerror directly (apparently strerror is not thread-safe).

I tried creating a buffer using strerrorlen_s but got the error: `error C3861: 'strerrorlen_s': identifier not found` and I saw [this](https://stackoverflow.com/questions/44430141/missing-c11-strerrorlen-s-function-under-msvc-2017) StackOverflow thread about it.

Reply:
Hint, save errno first before it gets overwritten by other functions.

Reply:
I don't think we need to be perfect here. strerrorlen_s is a nice to have but not required, so I think the simple buffer of 256 with sterror_s should be fine.